ghga_connector/core/main.py

```diff
@@ -186,37 +187,62 @@ def download(  # pylint: disable=too-many-arguments
         message_display.failure(f"The url {api_url} is currently not reachable.")
         raise exceptions.ApiNotReachableError(api_url=api_url)
 
-    public_key = base64.b64encode(crypt4gh.keys.get_public_key(pubkey_path)).decode()
+    public_key = base64.b64encode(crypt4gh.keys.get_public_key(pubkey_path))
+
+    # check output file
+    output_file = os.path.join(output_dir, file_id)
+    if os.path.isfile(output_file):
+        message_display.failure(f"The file {output_file} already exists.")
+        raise exceptions.FileAlreadyExistsError(output_file=output_file)
 
     # stage download and get file size
     download_url_tuple = await_download_url(
         api_url=api_url,
         file_id=file_id,
         max_wait_time=max_wait_time,
         message_display=message_display,
-        public_key=public_key,
     )
 
-    # perform the download:
+    # get file header envelope
+    try:
+        envelope = get_file_header_envelope(
+            file_id=file_id,
+            api_url=api_url,
+            public_key=public_key,
+        )
+    except (
+        exceptions.FileNotRegisteredError,
+        exceptions.EnvelopeNotFoundError,
+        exceptions.ExternalApiError,
+    ) as error:
+        message_display.failure(
+            f"The request to get an envelope for file {file_id} failed."
+        )
+        raise error
 
-    output_file = os.path.join(output_dir, file_id)
-    if os.path.isfile(output_file):
-        message_display.failure(f"The file {output_file} already exists.")
-        raise exceptions.FileAlreadyExistsError(output_file=output_file)
+    # put envelope in file
+    with open(output_file, "wb") as file:
+        file.write(envelope)
 
+    # perform the download
     try:
         download_parts(
             file_id=file_id,
             api_url=api_url,
             output_file=output_file,
             part_size=part_size,
             file_size=download_url_tuple[1],
-            public_key=public_key,
         )
     except exceptions.MaxRetriesReachedError as error:
         # Remove file, if the download failed.
         os.remove(output_file)
```

Comment:
Not sure if we need to have this if we open the file as above. "w" should truncate the file before writing.
If you want to keep it this way, there should also be cleanup in the next branch, i.e. envelope is already written before

Reply:
Added os.remove for both errors, also added append mode for the file, since the envelope by this point already exists.